src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -127,37 +126,22 @@ private async Task<BundleResponse> ExecuteTransactionForAllRequests(Hl7.Fhir.Mod
             return new BundleResponse(responseBundle.ToResourceElement());
         }
 
-        private static void ThrowTransactionException(HttpContext httpContext, OperationOutcome operationOutcome)
+        private async Task FillRequestLists(Hl7.Fhir.Model.Bundle.BundleType? bundleType, List<Hl7.Fhir.Model.Bundle.EntryComponent> bundleEntries)
         {
-            var operationOutcomeIssues = GetOperationOutcomeIssues(operationOutcome.Issue);
+            var resourceUrlList = new HashSet<string>();
 
-            var errorMessage = string.Format(Api.Resources.TransactionFailed, httpContext.Request.Method, httpContext.Request.Path);
-
-            throw new TransactionFailedException(errorMessage, (HttpStatusCode)httpContext.Response.StatusCode, operationOutcomeIssues);
-        }
-
-        private static List<OperationOutcomeIssue> GetOperationOutcomeIssues(List<OperationOutcome.IssueComponent> operationoutcomeIssueList)
-        {
-            var issues = new List<OperationOutcomeIssue>();
-
-            operationoutcomeIssueList.ForEach(x =>
-                issues.Add(new OperationOutcomeIssue(
-                    x.Severity.ToString(),
-                    x.Code.ToString(),
-                    x.Diagnostics)));
-
-            return issues;
-        }
-
-        private async Task FillRequestLists(List<Hl7.Fhir.Model.Bundle.EntryComponent> bundleEntries)
-        {
             foreach (Hl7.Fhir.Model.Bundle.EntryComponent entry in bundleEntries)
             {
                 if (entry.Request.Method == null)
                 {
                     continue;
                 }
 
+                if (bundleType == Hl7.Fhir.Model.Bundle.BundleType.Transaction)
+                {
+                    TransactionValidator.ValidateTransaction(resourceUrlList, entry);
```

Comment:
What is the advantage of having the caller keeping track of list of resource URLs instead of just pass the bundle in and let the function validates it upfront? For example, can you just pass the bundle to this ValidateTransaction (by the name, i think it needs to be renamed. It's not really validating the transaction.), and just have all of the logic encapsulated in that method?

At this point, the server hasn't replaced all of the ID yet so it seems like the detection is only applicable to resource IDs supplied by the user, in which case, it could just do the validation upfront and fail the request right away instead of allocating all these objects below that's just goint to be deallocated anyways?

Reply:
As discussed, will move validation as a separate module. For conditionacreate/update, will resolve the id's before executing the transaction.